sockeye/inference.py

```diff
@@ -1276,43 +1326,50 @@ def _beam_search(self,
+            pad_dist[:, C.PAD_ID] = mx.nd.where(mx.nd.clip(finished - inactive, 0, 1), scores_accumulated[:, 0], self.inf_array_long)
+            scores = mx.nd.where(finished + inactive, pad_dist, scores)
+
+            # (3) Get beam_size winning hypotheses for each sentence block separately. Only look as
+            # far as the active beam size for each sentence.
+            best_hyp_indices[:], best_word_indices[:], scores_accumulated[:, 0] = topk(scores)
 
             # Map from restricted to full vocab ids if needed
             if self.restrict_lexicon:
                 best_word_indices[:] = vocab_slice_ids.take(best_word_indices)
 
-            # (4) get hypotheses and their properties for beam_size winning hypotheses (ascending)
-            sequences = mx.nd.take(sequences, best_hyp_indices)
-            lengths = mx.nd.take(lengths, best_hyp_indices)
+            # (4) Normalize the scores of newly finished hypotheses. Note that after this until the
+            # next call to topk(), hypotheses may not be in sorted order.
             finished = mx.nd.take(finished, best_hyp_indices)
+            lengths = mx.nd.take(lengths, best_hyp_indices)
+            all_finished = ((best_word_indices == C.PAD_ID) + (best_word_indices == self.vocab_target[C.EOS_SYMBOL]))
+            newly_finished = all_finished - finished
+            scores_accumulated = mx.nd.where(newly_finished, scores_accumulated / self.length_penalty(lengths), scores_accumulated)
+            finished = all_finished
+            # All rows are now active (after special treatment of start state at t=1)
+            inactive[:] = 0
+
+            # (5) Prune out low-probability hypotheses. Pruning works by setting entries `inactive`.
+            if self.beam_prune > 0.0:
+                self._prune(scores_accumulated, best_word_indices, inactive, finished)
+
+            # (6) Update the beam with the hypotheses and their properties for the beam_size winning hypotheses (ascending)
+            sequences = mx.nd.take(sequences, best_hyp_indices)
             attention_scores = mx.nd.take(attention_scores, best_hyp_indices)
             attentions = mx.nd.take(attentions, best_hyp_indices)
 
-            # (5) update best hypotheses, their attention lists and lengths (only for non-finished hyps)
+            # (7) update best hypotheses, their attention lists and lengths (only for non-finished hyps)
             # pylint: disable=unsupported-assignment-operation
             sequences[:, t] = best_word_indices
             attentions[:, t, :] = attention_scores
             lengths += mx.nd.cast(1 - mx.nd.expand_dims(finished, axis=1), dtype='float32')
 
             # (6) optionally save beam history
             if self.store_beam:
-                unnormalized_scores = scores_accumulated * self.length_penalty(lengths - 1)
+                unnormalized_scores = mx.nd.where(finished, scores_accumulated / self.length_penalty(lengths), scores_accumulated)
```

Comment:
why normalize for `unnormalized_scores`? Why not just take `scores_accumulated` as is? `normalized_scores` down below should on the other hand now be normalized, right?

Reply:
Typo fixed (should be multiplying here, good catch). Normalization only applies to finished items now, so we only unnormalize there.

Reply:
oh ok, so you're unnormalizing the finished hypothesis. But then `beam_histories[sent]["normalized_scores"]` a little further down should be normalized, no?

Reply:
It is normalized, but only for finished hypotheses. And this has already been computed in `scores_accumulated` (line 1350).

Reply:
yes, so I'm wondering whether you should normalize these before adding them to `beam_histories[sent]["normalized_scores"]` to stay consistent with the behavior before this change. Something like:
```
normalized_scores = mx.nd.where(finished, scores_accumulated, scores_accumulated / self.length_penalty(lengths))
```
As this is only done when storing the beam the computational overhead should be fine.

Reply:
I have done this, but I don't know what the desired behavior is. Perhaps @Deseaus can comment?

Reply:
I think the visualization tools just report both the normalized and the unnormalized scores.